gdpr/docs/tunnistamo-kukkuu-configuration.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Culture Kids with Tunnistamo](#culture-kids-with-tunnistamo)
  - [OIDC provider clients](#oidc-provider-clients)
    - [Kukkuu API](#kukkuu-api)
    - [Kukkuu UI](#kukkuu-ui)
    - [Kukkuu Admin UI](#kukkuu-admin-ui)
  - [OIDC API](#oidc-api)
    - [Kukkuu API](#kukkuu-api-1)
  - [OIDC API scopes](#oidc-api-scopes)
    - [Kukkuu API Scope](#kukkuu-api-scope)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Culture Kids with Tunnistamo

## OIDC provider clients

Model: `oidc_provider.models.Client`

### Kukkuu API

Important fields with their values:

```python
{
  "name": "kukkuu-api",
  "response_types": [<ResponseType: code (Authorization Code Flow)>]>,
  "client_type": "public",
  "client_id": "https://api.hel.fi/auth/kukkuu",
  "client_secret": "",
  "jwt_alg": "RS256",
  "reuse_consent": True,
  "require_consent": True,
  "_redirect_uris": "http://localhost:8081/return http://localhost:3000/callback",
  "_scope": "openid kukkuu.gdprquery kukkuu.delete https://api.hel.fi/auth/kukkuu"
}
```

The following command can be used to create the client in Tunnistamo:

```shell
./manage.py add_oidc_client \
  --name "kukkuu-api" \
  --response_types "code" \
  --redirect_uris "http://localhost:8081/return http://localhost:3000/callback" \
  --client_id https://api.hel.fi/auth/kukkuu \
  --scopes "openid kukkuu.gdprquery kukkuu.delete https://api.hel.fi/auth/kukkuu" \
  --site_type dev \
  --login_methods "github"
```

> NOTE: this command does not set the consent setting.

### Kukkuu UI

Important fields with their values:

```python
{
  "name": "kukkuu-ui",
  "response_types": [<ResponseType: id_token token (Implicit Flow)>]>,
  "client_type": "public",
  "client_id": "https://api.hel.fi/auth/kukkuu-ui",
  "client_secret": "",
  "jwt_alg": "RS256",
  "reuse_consent": True,
  "require_consent": True,
  "_redirect_uris": "http://localhost:3000/callback http://localhost:3000/silent_renew.html http://localhost:3001/callback http://localhost:3001/silent_renew.html http://localhost:3002/callback http://localhost:3002/silent_renew.html",
  "_scope": "openid profile https://api.hel.fi/auth/kukkuu kukkuu.query kukkuu.delete https://api.hel.fi/auth/helsinkiprofile"
}
```

> NOTE: The Kukkuu UI has been using the implicit flow, but it probably needs to be changed when the Keycloak is taken in use.

The following command can be used to create the client in Tunnistamo:

```shell
./manage.py add_oidc_client \
  --name "kukkuu-ui" \
  --response_types "id_token token" \
  --redirect_uris "http://localhost:3000/callback http://localhost:3000/silent_renew.html http://localhost:3001/callback http://localhost:3001/silent_renew.html http://localhost:3002/callback http://localhost:3002/silent_renew.html" \
  --client_id https://api.hel.fi/auth/kukkuu-ui \
  --site_type dev \
  --login_methods "github"
```

> NOTE: this command does not set the scopes nor the consent setting.

### Kukkuu Admin UI

Important fields with their values:

```python
{
  "name": "kukkuu-admin-ui",
  "response_types": [<ResponseType: id_token token (Implicit Flow)>]>,
  "client_type": "public",
  "client_id": "https://api.hel.fi/auth/kukkuu-admin-ui",
  "client_secret": "",
  "jwt_alg": "RS256",
  "reuse_consent": True,
  "require_consent": True,
  "_redirect_uris": "http://localhost:3001/callback",
  "_scope": ""
}
```

> NOTE: The Kukkuu Admin UI has been using the implicit flow, but it probably needs to be changed when the Keycloak is taken in use.

The following command can be used to create the client in Tunnistamo:

```shell
./manage.py add_oidc_client \
  --name "kukkuu-admin-ui" \
  --response_types "id_token token" \
  --redirect_uris "http://localhost:3001/callback" \
  --client_id https://api.hel.fi/auth/kukkuu-admin-ui \
  --site_type dev \
  --login_methods "github"
```

> NOTE: this command does not set the scopes nor the consent setting.

## OIDC API

Model: `oidc_apis.models.Api`

### Kukkuu API

Important fields with their values:

```python
{
  'name': 'kukkuu',
  'required_scopes': ['email', 'profile'],
  'oidc_client': <Client: kukkuu-api>,
}
```

The following command can be used to create the API in Tunnistamo:

```shell
./manage.py add_oidc_api \
  --name kukkuu \
  --domain https://api.hel.fi/auth \
  --scopes profile email \
  --client_id https://api.hel.fi/auth/kukkuu
```

## OIDC API scopes

Model: `oidc_apis.models.ApiScope`

### Kukkuu API Scope

Important fields with their values:

```python
{
    "name": "kukkuu",
    "required_scopes": ["email", "profile"],
    "oidc_client": <Client: kukkuu-api>,
    "backchannel_logout_url": None
    "_allowed_apps": [<Client: kukkuu-admin-ui>, <Client: kukkuu-ui>]
    "scopes": [<ApiScope: https://api.hel.fi/auth/kukkuu.gdprdelete>, <ApiScope: https://api.hel.fi/auth/kukkuu.gdprquery>, <ApiScope: https://api.hel.fi/auth/kukkuu>]
}
```

The following command can be used to create the API scope in Tunnistamo:

```shell
./manage.py add_oidc_api_scope \
  --name "kukkuu" \
  --description "kukkuu" \
  --api_name kukkuu \
  --client_ids "https://api.hel.fi/auth/kukkuu"
```

The following command can be used to add the related clients to the created API scope:

```shell
./manage.py add_oidc_client_to_api_scope \
  -asi https://api.hel.fi/auth/kukkuu \
  -c https://api.hel.fi/auth/kukkuu-ui
./manage.py add_oidc_client_to_api_scope \
  -asi https://api.hel.fi/auth/kukkuu \
  -c https://api.hel.fi/auth/kukkuu-admin-ui
```


CHANGELOG.md
<!-- DOCTOC SKIP -->
<!-- REMINDER: While updating changelog, also remember to update
the version in kukkuu/__init.py__ -->

## [3.11.0](https://github.com/City-of-Helsinki/kukkuu/compare/kukkuu-v3.10.0...kukkuu-v3.11.0) (2025-01-15)


### Features

* **auditlog-extra:** Add request path to additional data in LogEntry ([59c3d46](https://github.com/City-of-Helsinki/kukkuu/commit/59c3d4685e65808a399d7b58d85280d7e35145de))
* **auditlog-extra:** Admin mixin ([8f4e56a](https://github.com/City-of-Helsinki/kukkuu/commit/8f4e56a4f394769741f6d61607c1e5a36bd8d550))
* **auditlog-extra:** Configuration helper ([00165e6](https://github.com/City-of-Helsinki/kukkuu/commit/00165e6ad89bf0a1bca6f83b55d65597808c0cde))
* **auditlog-extra:** Decorator ([e7b9257](https://github.com/City-of-Helsinki/kukkuu/commit/e7b92579e3fc08e5ac6ea7b851bb144a8e3b6bad))
* **auditlog-extra:** Middleware ([0e63451](https://github.com/City-of-Helsinki/kukkuu/commit/0e63451fc9771c7c39d81bb5b6e658e9e353407b))
* **auditlog:** Installed django-auditlog ([e0f2f04](https://github.com/City-of-Helsinki/kukkuu/commit/e0f2f04b837d9561576659b67304fe043db7dbd7))
* **auditlog:** Write access log of restricted graphene nodes ([9ba8fad](https://github.com/City-of-Helsinki/kukkuu/commit/9ba8fad2c93d1454500d1d3eb2864c2739aa6786))
* **auditlog:** Write access logs of restricted admin models ([e4fa72d](https://github.com/City-of-Helsinki/kukkuu/commit/e4fa72daa683d73dc7f1a4c29bbb8c2a39765385))
* **csp:** Install and configure django-csp ([8c6bb7d](https://github.com/City-of-Helsinki/kukkuu/commit/8c6bb7d5f8b251da5cd048630f273a709e9950dc))
* **devops:** Liveness probe to test the database connection ([1dcc0ff](https://github.com/City-of-Helsinki/kukkuu/commit/1dcc0ff2ba76819fd3e8235da0a07e622a573c1b))
* **devops:** Readiness probe to give build and release info ([f8db348](https://github.com/City-of-Helsinki/kukkuu/commit/f8db34822e3d94ef85f0ca112354982d48abc3a8))
* **sentry:** Ignore authentication expired errors in Sentry ([f232111](https://github.com/City-of-Helsinki/kukkuu/commit/f2321118d2d691b95a7ca268e1d03d2bb18a5269))


### Bug Fixes

* "The USE_L10N setting is deprecated" deprecation warning ([9f0abc1](https://github.com/City-of-Helsinki/kukkuu/commit/9f0abc17873af8d26657c7abf64a7628701ce328))
* /graphql/ GraphiQL interface by changing CSP rules ([0a01103](https://github.com/City-of-Helsinki/kukkuu/commit/0a011035749834e0b9bcf7dd48a0790064889b79))
* **auditlog-extra:** Admin mixin should use get_changelist_instance ([7d71f4c](https://github.com/City-of-Helsinki/kukkuu/commit/7d71f4c94ce7ee829880c726c5caf19c4d41d870))
* **auditlog-extra:** Duplicated log creation prevention in admin mixin ([a427992](https://github.com/City-of-Helsinki/kukkuu/commit/a427992e21f969825f520badd426a646e84b320f))
* **auditlog-extra:** Explicitly define app_label for DummyTestModel ([551eb46](https://github.com/City-of-Helsinki/kukkuu/commit/551eb4616c65d3b426317bffa4f3b20f337bd89b))
* **auditlog:** Exclude mailer models from auditlog ([3f06c0b](https://github.com/City-of-Helsinki/kukkuu/commit/3f06c0bb503854d582b80d9b535393f236f288fc))
* **config:** Admin ui url ([f5e0d24](https://github.com/City-of-Helsinki/kukkuu/commit/f5e0d2418516e9745d06c9c1992af7f6a502ff71))

kukkuu/settings.py
import os
import subprocess
from datetime import datetime

import environ
import sentry_sdk
from django.core.exceptions import ImproperlyConfigured
from django.utils.translation import gettext_lazy as _
from jose import ExpiredSignatureError
from sentry_sdk.integrations.django import DjangoIntegration
from sentry_sdk.types import Event, Hint

from kukkuu.consts import CSP
from kukkuu.exceptions import AuthenticationExpiredError
from kukkuu.tests.utils.jwt_utils import is_valid_256_bit_key

checkout_dir = environ.Path(__file__) - 2
assert os.path.exists(checkout_dir("manage.py"))

parent_dir = checkout_dir.path("..")
if os.path.isdir(parent_dir("etc")):
    env_file = parent_dir("etc/env")
    default_var_root = environ.Path(parent_dir("var"))
else:
    env_file = checkout_dir(".env")
    default_var_root = environ.Path(checkout_dir("var"))

env = environ.Env(
    DEBUG=(bool, False),
    SECRET_KEY=(str, ""),
    MEDIA_ROOT=(environ.Path(), environ.Path(checkout_dir("var"))("media")),
    STATIC_ROOT=(environ.Path(), default_var_root("static")),
    MEDIA_URL=(str, "/media/"),
    STATIC_URL=(str, "/static/"),
    ALLOWED_HOSTS=(list, []),
    USE_X_FORWARDED_HOST=(bool, False),
    DATABASE_URL=(str, "postgres://kukkuu:kukkuu@localhost/kukkuu"),
    CACHE_URL=(str, "locmemcache://"),
    MAILER_EMAIL_BACKEND=(str, "django.core.mail.backends.console.EmailBackend"),
    MAILER_LOCK_PATH=(str, "/tmp/mailer_lockfile"),
    DEFAULT_FROM_EMAIL=(str, "kukkuu@example.com"),
    DEFAULT_SMS_SENDER=(str, "Hel.fi"),
    ILMOITIN_TRANSLATED_FROM_EMAIL=(dict, {}),
    TRANSLATED_SMS_SENDER=(dict, {}),
    MAIL_MAILGUN_KEY=(str, ""),
    MAIL_MAILGUN_DOMAIN=(str, ""),
    MAIL_MAILGUN_API=(str, ""),
    NOTIFICATION_SERVICE_API_TOKEN=(str, ""),
    NOTIFICATION_SERVICE_API_URL=(str, "https://notification-service.hel.fi/v1/"),
    SENTRY_DSN=(str, ""),
    SENTRY_ENVIRONMENT=(str, ""),
    CORS_ALLOWED_ORIGINS=(list, []),
    CORS_ALLOWED_ORIGIN_REGEXES=(list, []),
    CORS_ORIGIN_ALLOW_ALL=(bool, False),
    TOKEN_AUTH_ACCEPTED_SCOPE_PREFIX=(str, ""),
    TOKEN_AUTH_REQUIRE_SCOPE_PREFIX=(bool, False),
    TOKEN_AUTH_ACCEPTED_AUDIENCE=(list, ["https://api.hel.fi/auth/kukkuu"]),
    TOKEN_AUTH_AUTHSERVER_URL=(
        list,
        ["https://tunnistus.test.hel.ninja/auth/realms/helsinkitunnistus"],
    ),
    ILMOITIN_QUEUE_NOTIFICATIONS=(bool, True),
    STORAGES_DEFAULT_BACKEND=(str, "django.core.files.storage.FileSystemStorage"),
    GS_BUCKET_NAME=(str, ""),
    STAGING_GCS_BUCKET_CREDENTIALS=(str, ""),
    GS_DEFAULT_ACL=(str, "publicRead"),
    GS_FILE_OVERWRITE=(bool, False),
    AZURE_ACCOUNT_NAME=(str, ""),
    AZURE_ACCOUNT_KEY=(str, ""),
    AZURE_CONTAINER=(str, ""),
    AZURE_BLOB_STORAGE_SAS_TOKEN=(str, ""),
    ENABLE_GRAPHIQL=(bool, False),
    KUKKUU_UI_BASE_URL=(str, "http://localhost:3000"),
    KUKKUU_TICKET_VERIFICATION_URL=(str, ""),
    KUKKUU_HASHID_SALT=(str, None),
    KUKKUU_ENROLLED_OCCURRENCE_IN_PAST_LEEWAY=(int, 60),
    KUKKUU_DEFAULT_EVENT_DURATION=(int, 120),
    KUKKUU_REMINDER_DAYS_IN_ADVANCE=(int, 1),
    KUKKUU_FEEDBACK_NOTIFICATION_DELAY=(int, 15),
    KUKKUU_NOTIFICATIONS_SHEET_ID=(str, ""),
    VERIFICATION_TOKEN_VALID_MINUTES=(int, 15),
    VERIFICATION_TOKEN_LENGTH=(int, 8),
    SUBSCRIPTIONS_AUTH_TOKEN_VALID_MINUTES=(int, 30 * 24 * 60),  # 30 days
    SUBSCRIPTIONS_AUTH_TOKEN_LENGTH=(int, 16),
    GDPR_API_QUERY_SCOPE=(str, "gdprquery"),
    GDPR_API_DELETE_SCOPE=(str, "gdprdelete"),
    GDPR_API_AUTHORIZATION_FIELD=(str, "authorization.permissions.scopes"),
    HELUSERS_BACK_CHANNEL_LOGOUT_ENABLED=(bool, False),
    TOKEN_AUTH_BROWSER_TEST_JWT_256BIT_SIGN_SECRET=(str, None),
    TOKEN_AUTH_BROWSER_TEST_JWT_ISSUER=(list, None),
    TOKEN_AUTH_BROWSER_TEST_ENABLED=(bool, False),
    BROWSER_TEST_PROJECT_YEAR=(int, 1234),
    BROWSER_TEST_PROJECT_NAME=(str, "Browser test"),
    BROWSER_TEST_GROUP_NAME=(str, "Browser test"),
    BROWSER_TEST_AD_GROUP_NAME=(str, "kukkuu_browser_test"),
    KUKKUU_DEFAULT_LOGGING_LEVEL=(str, "INFO"),
    APP_RELEASE=(str, ""),
)

if os.path.exists(env_file):
    env.read_env(env_file)

BASE_DIR = str(checkout_dir)

DEBUG = env.bool("DEBUG")
SECRET_KEY = env.str("SECRET_KEY")
if DEBUG and not SECRET_KEY:
    SECRET_KEY = "xxx"

ALLOWED_HOSTS = env.list("ALLOWED_HOSTS")
USE_X_FORWARDED_HOST = env.bool("USE_X_FORWARDED_HOST")

DATABASES = {"default": env.db()}

CACHES = {"default": env.cache()}

if env.str("DEFAULT_FROM_EMAIL"):
    DEFAULT_FROM_EMAIL = env.str("DEFAULT_FROM_EMAIL")
if env("MAIL_MAILGUN_KEY"):
    ANYMAIL = {
        "MAILGUN_API_KEY": env("MAIL_MAILGUN_KEY"),
        "MAILGUN_SENDER_DOMAIN": env("MAIL_MAILGUN_DOMAIN"),
        "MAILGUN_API_URL": env("MAIL_MAILGUN_API"),
    }
EMAIL_BACKEND = "mailer.backend.DbBackend"
MAILER_EMAIL_BACKEND = env.str("MAILER_EMAIL_BACKEND")
if env("MAILER_LOCK_PATH"):
    MAILER_LOCK_PATH = env.str("MAILER_LOCK_PATH")

ILMOITIN_TRANSLATED_FROM_EMAIL = env("ILMOITIN_TRANSLATED_FROM_EMAIL")
ILMOITIN_QUEUE_NOTIFICATIONS = env("ILMOITIN_QUEUE_NOTIFICATIONS")

if env("DEFAULT_SMS_SENDER"):
    DEFAULT_SMS_SENDER = env.str("DEFAULT_SMS_SENDER")

TRANSLATED_SMS_SENDER = env("TRANSLATED_SMS_SENDER")
NOTIFICATION_SERVICE_API_TOKEN = env.str("NOTIFICATION_SERVICE_API_TOKEN")
NOTIFICATION_SERVICE_API_URL = env.str("NOTIFICATION_SERVICE_API_URL")

try:
    REVISION = (
        subprocess.check_output(["git", "rev-parse", "--short", "HEAD"])
        .strip()
        .decode("utf-8")
    )
except Exception:
    REVISION = "n/a"


def sentry_before_send(event: Event, hint: Hint):
    """
    Filter out uninformative errors from being sent to Sentry.

    Some Graphene errors are generic Exceptions with unhelpful messages.
    This function filters them out to reduce noise in Sentry.

    NOTE: The `ignore_errors` option is actually deprecated,
    nevertheless it's included in the init configuration.
    See https://github.com/getsentry/sentry-python/issues/149

    Args:
        event: The event that is being sent to Sentry.
        hint: A dictionary containing additional information about the event.

    Returns:
        The event if it should be sent to Sentry, or None if it should be ignored.
    """

    IGNORED_ERRORS_CLASSES = (ExpiredSignatureError, AuthenticationExpiredError)
    if "exc_info" in hint:
        exc_type, exc_value, traceback = hint["exc_info"]
        if isinstance(exc_value, IGNORED_ERRORS_CLASSES):
            return None
    return event


sentry_sdk.init(
    dsn=env.str("SENTRY_DSN"),
    release=REVISION,
    environment=env("SENTRY_ENVIRONMENT"),
    integrations=[DjangoIntegration()],
    before_send=sentry_before_send,
)
sentry_sdk.integrations.logging.ignore_logger("graphql.execution.utils")

MEDIA_ROOT = env("MEDIA_ROOT")
STATIC_ROOT = env("STATIC_ROOT")
MEDIA_URL = env.str("MEDIA_URL")
STATIC_URL = env.str("STATIC_URL")

# For staging env, we use Google Cloud Storage
STORAGES_DEFAULT_BACKEND = env("STORAGES_DEFAULT_BACKEND")

# For prod, it's Azure Storage
if STORAGES_DEFAULT_BACKEND == "storages.backends.azure_storage.AzureStorage":
    AZURE_ACCOUNT_NAME = env("AZURE_ACCOUNT_NAME")
    AZURE_CONTAINER = env("AZURE_CONTAINER")
    if env("AZURE_BLOB_STORAGE_SAS_TOKEN"):
        SAS_TOKEN = env("AZURE_BLOB_STORAGE_SAS_TOKEN")
        AZURE_ENDP = f"BlobEndpoint=https://{AZURE_ACCOUNT_NAME}.blob.core.windows.net"
        AZURE_CONNECTION_STRING = f"{AZURE_ENDP};SharedAccessSignature={SAS_TOKEN};"
    else:
        AZURE_ACCOUNT_KEY = env("AZURE_ACCOUNT_KEY")

STORAGES = {
    "default": {
        "BACKEND": STORAGES_DEFAULT_BACKEND,
    },
    "staticfiles": {
        "BACKEND": "django.contrib.staticfiles.storage.StaticFilesStorage",
    },
}

ROOT_URLCONF = "kukkuu.urls"
WSGI_APPLICATION = "kukkuu.wsgi.application"

LANGUAGE_CODE = "fi"
LANGUAGES = (("fi", _("Finnish")), ("en", _("English")), ("sv", _("Swedish")))
TIME_ZONE = "Europe/Helsinki"
USE_I18N = True
USE_TZ = True
# Set to True to enable GraphiQL interface, this will overriden to True if DEBUG=True
ENABLE_GRAPHIQL = env("ENABLE_GRAPHIQL")

INSTALLED_APPS = [
    "helusers.apps.HelusersConfig",
    "helusers.apps.HelusersAdminConfig",
    "kukkuu_helusers_admin",  # must be after `helusers`, since it overrides admin
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "corsheaders",
    "csp",
    "graphene_django",
    "parler",
    "anymail",
    "mailer",
    "kukkuu_mailer_admin",  # must be after `mailer`, since it overrides admin
    "django_ilmoitin",
    "django_filters",
    "auditlog",
    "guardian",
    "rest_framework",
    "drf_spectacular",
    "health_check",  # requirement
    # local apps
    "custom_health_checks",
    "users",
    "children",
    "utils",
    "projects",
    "events",
    "venues",
    "languages",
    "subscriptions",
    "messaging",
    "importers",
    "reports",
    "verification_tokens",
    "hel_django_auditlog_extra.apps.HelDjangoAuditLogExtraConfig",
    "kukkuu",
    "django_cleanup.apps.CleanupConfig",  # This must be included last
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "csp.middleware.CSPMiddleware",
    "django.middleware.locale.LocaleMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "hel_django_auditlog_extra.middleware.AuditlogMiddleware",
]

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ]
        },
    }
]

CORS_ALLOWED_ORIGINS = env.list("CORS_ALLOWED_ORIGINS")
CORS_ALLOWED_ORIGIN_REGEXES = env.list("CORS_ALLOWED_ORIGIN_REGEXES")
CORS_ORIGIN_ALLOW_ALL = env.bool("CORS_ORIGIN_ALLOW_ALL")

# Configure the default CSP rule for different source types
CSP_DEFAULT_SRC = ["'self'"]

# CSP_STYLE_SRC includes 'unsafe-inline' for inline styles added by `django-helusers`
# and `SpectacularRedocView`.
CSP_STYLE_SRC = [
    CSP.SELF,
    CSP.UNSAFE_INLINE,
    "cdn.jsdelivr.net",
    "fonts.googleapis.com",
]

# CSP_SCRIPT_SRC includes 'unsafe-eval' for inline scripts added by
# `SpectacularRedocView`
CSP_SCRIPT_SRC = [
    CSP.SELF,
    CSP.UNSAFE_EVAL,
    "cdn.jsdelivr.net",
    "blob:",
]

# CSP_FONT_SRC includes Google fonts fetched by `SpectacularRedocView`
CSP_FONT_SRC = [
    CSP.SELF,
    "fonts.googleapis.com",
    "fonts.gstatic.com",
    "data:",  # /graphql/ endpoint uses "data:font/woff2"
]

# CSP_IMG_SRC includes sources for images fetched by `SpectacularRedocView`
CSP_IMG_SRC = [
    CSP.SELF,
    "cdn.redoc.ly",
    "blob:",
    "data:",
]

AUTH_USER_MODEL = "users.User"

AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
    "kukkuu.oidc.BrowserTestAwareJWTAuthentication",
    "guardian.backends.ObjectPermissionBackend",
]

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

ANONYMOUS_USER_NAME = None  # we don't need django-guardian's AnonymousUser

OIDC_API_TOKEN_AUTH = {
    "AUDIENCE": env.list("TOKEN_AUTH_ACCEPTED_AUDIENCE"),
    "API_SCOPE_PREFIX": env.str("TOKEN_AUTH_ACCEPTED_SCOPE_PREFIX"),
    "ISSUER": env.list("TOKEN_AUTH_AUTHSERVER_URL"),
    "REQUIRE_API_SCOPE_FOR_AUTHENTICATION": env.bool("TOKEN_AUTH_REQUIRE_SCOPE_PREFIX"),
    "API_AUTHORIZATION_FIELD": env.str("GDPR_API_AUTHORIZATION_FIELD"),
}

OIDC_AUTH = {"OIDC_LEEWAY": 60 * 60}

OIDC_BROWSER_TEST_API_TOKEN_AUTH = {
    "ENABLED": env.bool("TOKEN_AUTH_BROWSER_TEST_ENABLED"),
    "JWT_SIGN_SECRET": env.str("TOKEN_AUTH_BROWSER_TEST_JWT_256BIT_SIGN_SECRET"),
    "ISSUER": env.list("TOKEN_AUTH_BROWSER_TEST_JWT_ISSUER"),
    "AUDIENCE": env.list("TOKEN_AUTH_ACCEPTED_AUDIENCE"),
    "API_SCOPE_PREFIX": env.str("TOKEN_AUTH_ACCEPTED_SCOPE_PREFIX"),
    "REQUIRE_API_SCOPE_FOR_AUTHENTICATION": env.bool("TOKEN_AUTH_REQUIRE_SCOPE_PREFIX"),
    "API_AUTHORIZATION_FIELD": env.str("GDPR_API_AUTHORIZATION_FIELD"),
}

# Ensure that the browser test JWT authentication is configured properly.
if OIDC_BROWSER_TEST_API_TOKEN_AUTH["ENABLED"]:
    if not OIDC_BROWSER_TEST_API_TOKEN_AUTH["ISSUER"]:
        raise ImproperlyConfigured(
            "API token authentication is enabled, but no issuer is configured. "
            "Set OIDC_BROWSER_TEST_API_TOKEN_AUTH['ISSUER']."
        )
    if not is_valid_256_bit_key(OIDC_BROWSER_TEST_API_TOKEN_AUTH["JWT_SIGN_SECRET"]):
        raise ImproperlyConfigured(
            "JWT secret key for BrowserTestAwareJWTAuthentication must be 256-bits. "
            "Set OIDC_BROWSER_TEST_API_TOKEN_AUTH['JWT_SIGN_SECRET']."
        )

# The browser tests should be targeted for the users of this group.
# If the group does not exist, it will be automatically created,
# when the OIDC_BROWSER_TEST_API_TOKEN_AUTH["ENABLED"] is True.
BROWSER_TEST_PROJECT_YEAR = env.int("BROWSER_TEST_PROJECT_YEAR")
BROWSER_TEST_PROJECT_NAME = env.str("BROWSER_TEST_PROJECT_NAME")
BROWSER_TEST_GROUP_NAME = env.str("BROWSER_TEST_GROUP_NAME") or env.str(
    "BROWSER_TEST_PROJECT_NAME"
)
BROWSER_TEST_AD_GROUP_NAME = env.str("BROWSER_TEST_AD_GROUP_NAME")

SITE_ID = 1

PARLER_LANGUAGES = {SITE_ID: ({"code": "fi"}, {"code": "sv"}, {"code": "en"})}

PARLER_SUPPORTED_LANGUAGE_CODES = [x["code"] for x in PARLER_LANGUAGES[SITE_ID]]

PARLER_ENABLE_CACHING = False

GRAPHENE = {
    "SCHEMA": "kukkuu.schema.schema",
    "MIDDLEWARE": [
        "graphql_jwt.middleware.JSONWebTokenMiddleware",
    ],
}

GRAPHQL_JWT = {"JWT_AUTH_HEADER_PREFIX": "Bearer"}

KUKKUU_MAX_NUM_OF_CHILDREN_PER_GUARDIAN = 100
KUKKUU_QUERY_MAX_DEPTH = 12
KUKKUU_UI_BASE_URL = env("KUKKUU_UI_BASE_URL")
# IF KUKKUU_TICKET_VERIFICATION_URL is set to None,
# the qr code won't be attached to the enrolment notification email.
# Use {reference_id} as a specified value in the given string
# e.g http://localhost:3000/check-ticket-validity/{reference_id}
KUKKUU_TICKET_VERIFICATION_URL = env("KUKKUU_TICKET_VERIFICATION_URL")
# How much an enrolled occurrence can be in the past and still be considered as
# not being in the past. In minutes.
KUKKUU_ENROLLED_OCCURRENCE_IN_PAST_LEEWAY = env(
    "KUKKUU_ENROLLED_OCCURRENCE_IN_PAST_LEEWAY"
)
# If event duration is missing, this will be used as a default value. In minutes.
KUKKUU_DEFAULT_EVENT_DURATION = env("KUKKUU_DEFAULT_EVENT_DURATION")
KUKKUU_DEFAULT_ENROLMENT_LIMIT = 2
KUKKUU_REMINDER_DAYS_IN_ADVANCE = env("KUKKUU_REMINDER_DAYS_IN_ADVANCE")
KUKKUU_FEEDBACK_NOTIFICATION_DELAY = env("KUKKUU_FEEDBACK_NOTIFICATION_DELAY")
KUKKUU_NOTIFICATIONS_SHEET_ID = env("KUKKUU_NOTIFICATIONS_SHEET_ID")

KUKKUU_HASHID_MIN_LENGTH = 5
KUKKUU_HASHID_ALPHABET = "abcdefghijklmnopqrstuvwxyz"
KUKKUU_HASHID_SALT = env.str("KUKKUU_HASHID_SALT")
if KUKKUU_HASHID_SALT is None:
    raise ImproperlyConfigured(
        "KUKKUU_HASHID_SALT must be configured! "
        "Hashids in Kukkuu utils requires the setting!"
    )


REST_FRAMEWORK = {
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "rest_framework.authentication.BasicAuthentication"
    ]
    + (["rest_framework.authentication.SessionAuthentication"] if DEBUG else []),
    "DEFAULT_PERMISSION_CLASSES": ["reports.drf_permissions.ReportAPIPermission"],
    "DEFAULT_RENDERER_CLASSES": ["rest_framework.renderers.JSONRenderer"]
    + (["rest_framework.renderers.BrowsableAPIRenderer"] if DEBUG else []),
    "TEST_REQUEST_DEFAULT_FORMAT": "json",
    "DEFAULT_SCHEMA_CLASS": "drf_spectacular.openapi.AutoSchema",
}

SPECTACULAR_SETTINGS = {"TITLE": "Kukkuu report API", "VERSION": "1.0.0"}

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "timestamped_named": {
            "format": "%(asctime)s %(name)s %(levelname)s: %(message)s"
        }
    },
    "handlers": {
        "console": {"class": "logging.StreamHandler", "formatter": "timestamped_named"}
    },
    "loggers": {
        "": {
            "handlers": ["console"],
            "level": env("KUKKUU_DEFAULT_LOGGING_LEVEL", None, "INFO"),
        }
    },
}

# GDPR API DOCS - See https://profile-api.dev.hel.ninja/docs/gdpr-api/.
GDPR_API_MODEL = "users.User"
GDPR_API_QUERY_SCOPE = env("GDPR_API_QUERY_SCOPE")
GDPR_API_DELETE_SCOPE = env("GDPR_API_DELETE_SCOPE")
GDPR_API_MODEL_LOOKUP = "uuid"
GDPR_API_URL_PATTERN = "v1/user/<uuid:uuid>"
GDPR_API_USER_PROVIDER = "gdpr.service.get_user"
GDPR_API_DELETER = "gdpr.service.clear_data"

HELUSERS_BACK_CHANNEL_LOGOUT_ENABLED = env("HELUSERS_BACK_CHANNEL_LOGOUT_ENABLED")

# release information
APP_RELEASE = env("APP_RELEASE")
# get build time from a file in docker image
APP_BUILD_TIME = datetime.fromtimestamp(os.path.getmtime(__file__))

# Load auditlog settings
from kukkuu.auditlog_settings import *  # noqa: E402, F403

# local_settings.py can be used to override environment-specific settings
# like database and email that differ between development and production.
local_settings_path = os.path.join(checkout_dir(), "local_settings.py")
if os.path.exists(local_settings_path):
    with open(local_settings_path) as fp:
        code = compile(fp.read(), local_settings_path, "exec")
    exec(code, globals(), locals())
